sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala

```diff
@@ -469,4 +470,23 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
 
     checkAnswer(sql("""select json[0] from jsonTable"""), Seq(Row(null)))
   }
+
+  test("from_json invalid json - check modes") {
+    val df = Seq("""{"a" 1}""", """{"a": 2}""").toDS()
+    val schema = new StructType().add("a", IntegerType)
+
+    checkAnswer(
+      df.select(from_json($"value", schema, Map("mode" -> "PERMISSIVE"))),
+      Row(Row(null)) :: Row(Row(2)) :: Nil)
+
+    val exceptionOne = intercept[SparkException] {
+      df.select(from_json($"value", schema, Map("mode" -> "FAILFAST"))).collect()
```

Comment:
`JsonToStructs` resembles PERMISSIVE mode (from the first place) although their behaviours are slightly different. This is going to be different with PERMISSIVE and also FAILFAST modes. They are actually behaviour changes if we just use PERMISSIVE mode here by default (as @viirya pointed out).

Reply:
Behavior of `JsonToStructs` is pretty close to `PERMISSIVE` actually. I have to make just a few small changes in tests that checks processing malformed inputs.